Add notice when premium features for key are available [MAILPOET-5429]

lysyjan opened the pull request:
## Description

To improve our user experience, this pull request adds a new notice that should be displayed only when the customer's key supports the premium feature and the premium plugin is not installed or downloaded.

## Code review notes

_N/A_

## QA notes

1. Set a key supporting premium features
  a. Check if the new notice is visible when the premium plugin is not installed. The message should contain a download link. As a part of it, the exact text should be visible in the premium banner, which is visible on the newsletter stats page.
  b. Check if the new notice is visible when the premium plugin is not activated. The message should contain an activation link. As a part of it, the exact text should be visible in the premium banner, which is visible on the newsletter stats page.
2. Use a key without premium features; the new notices shouldn't be visible

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5429]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5429]: https://mailpoet.atlassian.net/browse/MAILPOET-5429?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

lysyjan committed:
Add notice when premium features for key are available

[MAILPOET-5429]

lysyjan committed:
Improve rendering premium banner when premium plugin is downloaded or inactive

[MAILPOET-5429]

lysyjan committed:
Refactor premium banner to tsx

[MAILPOET-5429]

johnolek commented on mailpoet/lib/Util/Notices/PremiumFeaturesAvailableNotice.php:
```diff
@@ -0,0 +1,76 @@
+    $noticeString = Helpers::replaceLinkTags($noticeString, $link, $attributes);
+    $extraClasses = 'mailpoet-dismissible-notice is-dismissible';
+
+    return Notice::displayInfo($noticeString, $extraClasses, self::OPTION_NAME);
```
The specs say to `display a green notice`, so should we use `Notice::displaySuccess` here?

johnolek commented on mailpoet/assets/js/src/newsletters/campaign_stats/premium_banner.jsx:
```diff
@@ -5,33 +5,69 @@ import { PremiumRequired } from 'common/premium_required/premium_required';
+        )}
+      </p>
+    );
+    ctaButton = (
+      <Button
+        href={MailPoet.premiumPluginActivationUrl}
+        rel="noopener noreferrer"
+      >
+        {__('Activate MailPoet Premium plugin', 'mailpoet')}
+      </Button>
+    );
+    // If the premium plugin is not installed, we need to provide a download link
+    if (!MailPoet.isPremiumPluginInstalled) {
+      ctaButton = (
+        <Button
+          href={MailPoet.premiumPluginDownloadUrl}
+          target="_blank"
+          rel="noopener noreferrer"
+        >
+          {__('Download MailPoet Premium plugin', 'mailpoet')}
+        </Button>
+      );
+    }
```
This looks good to me as is, but I was wondering what you think about turning it into an ajax request instead so the user doesn't have to hit the back button themselves. ChatGPT helped me come up with this solution, which could definitely be improved but seems to work:

```js
function SkipDisplayingDetailedStats() {
  let ctaButton;
  let description;

  const [loading, setLoading] = useState(false);

  if (
    MailPoet.hasValidPremiumKey &&
    (!MailPoet.isPremiumPluginInstalled || !MailPoet.premiumActive)
  ) {
    description = (
      <p>
        {__(
          'Your current MailPoet plan includes advanced features, but they require the MailPoet Premium plugin to be installed and activated.',
          'mailpoet',
        )}
      </p>
    );
    ctaButton = (
      <Button
        withSpinner={loading}
        href={MailPoet.premiumPluginActivationUrl}
        rel="noopener noreferrer"
        onClick={(e) => {
          e.preventDefault();
          setLoading(true);

          jQuery.get(MailPoet.premiumPluginActivationUrl, (response) => {
            if (response.includes('Plugin activated')) {
              window.location.reload();
            } else {
              setLoading(false);
              console.error('Error activating plugin:', response);
            }
          });
        }}
        disabled={loading}
      >
        {loading
          ? __('Activating MailPoet premium...', 'mailpoet')
          : __('Activate MailPoet Premium plugin', 'mailpoet')}
      </Button>
    );
```

I'm totally happy to leave it as is though!

lysyjan replied:
@johnolek I really like your suggestion 👍 I will change it.

johnolek reviewed (dismissed):
Looks good to me! I had one idea for your consideration and also a question about the notice style. Neither should be considered required changes.

lysyjan committed:
Change info notice to success notice

[MAILPOET-5429]

lysyjan committed:
Use ajax call for the premium plugin activation in the premium banner

[MAILPOET-5429]

lysyjan commented:
@johnolek I addressed your notes in the last two commits. Also, I rebased this branch on the latest trunk, I you are okay with that.
Because `console.error` is not allowed by our ESLint configuration, the error state of the Ajax call is not handled, and I didn't have any idea how can I do it without significant changes. Of course, I'm all ears if you have some ideas.

johnolek commented:
@lysyjan Thanks for the updates!

For the error case I would be happy if we just include some kind of user facing message, the easiest way probably being `MailPoet.Notice.error`. Something along the lines of `We were unable to activate the premium plugin, please try visiting the [plugin page link] to activate it manually.`

lysyjan committed:
Add better handling of a failed ajax request

[MAILPOET-5429]

johnolek reviewed (dismissed):
Looks good to me! And thanks for being open to the different idea

Aschepikov commented:
HI @lysyjan with a free key on the segment page when I am adding the second filter I can see the message "download mailpoet premium plugin" 
<img width="1462" alt="image" src="https://github.com/mailpoet/mailpoet/assets/113185437/86cfde49-014c-49dc-be2b-a78bb6fbc530">
on other premium features, there are upgrade messages 
![image](https://github.com/mailpoet/mailpoet/assets/113185437/65417ecd-c71f-4530-bf46-ce43ee08fa47)
Can you check it please

lysyjan committed:
Reuse existing component and fixed behavior checking premium accessibility

[MAILPOET-5429]

lysyjan committed:
Use Subscribers feature instead of Installer

[MAILPOET-5429]

lysyjan commented:
@Aschepikov Nice catch. This issue was separate from my changes. Anyway, I fixed this behavior on my PR.

@johnolek Please, take a look at the two latest commits.
- https://github.com/mailpoet/mailpoet/pull/5059/commits/3987ebb05aef69cc3d527b930c8ff337d0a176f3 This commit fixes an issue found by @Aschepikov. I decided to revert my changes and moved them to a higher level, where a part of this logic already was. The code that we discussed in this PR (activation with Ajax call).
- https://github.com/mailpoet/mailpoet/pull/5059/commits/8d93f218a46a9f201ee9c0720b7698fc0f5e006b The second commit unifies the behavior with the Segment page.

johnolek reviewed (approved):
The latest changes look good to me 👍

Aschepikov reviewed (approved):
GJ!
